jstz_api/src/encoding/text_decoder.rs

```diff
@@ -0,0 +1,381 @@
+        self.io_queue.drain(0..read);
+    }
+
+    //  https://encoding.spec.whatwg.org/#dom-textdecoder-decode
+    fn decode(
+        // input: Option<allowsharedbufersource>,
+        &mut self,
+        input: Option<Vec<u8>>,
+        options: Option<TextDecodeOptions>,
+    ) -> JsResult<String> {
+        //  Handle optional parameters
+        let input = input.unwrap_or(vec![]);
+        let options = options.unwrap_or(TextDecodeOptions::default());
+        //  1. If this's do not flush is false,
+        if !self.do_not_flush {
+            //  then set this's decoder to a new instance of this's encoding's decoder,
+            self.decoder = if self.ignore_bom {
+                self.encoding.new_decoder_without_bom_handling()
+            } else {
+                self.encoding.new_decoder()
+            };
+            //  this's I/O queue to the I/O queue of bytes <<end-of-queue>>,
+            self.io_queue.clear();
+            //  and this's BOM seen to false.
+            //  NB: we do not need this flag (intern to `encoding_rs` crate decode functions)
+        };
+        //  2. Set this's do not flush to options["stream"].
+        self.do_not_flush = options.stream;
+        //  3. If input is given, then push a copy of input to this's I/O queue.
+        //  FIXME: map this error into a JsError ? which one ?
+        let _: Result<usize, std::io::Error> = self.io_queue.write(input.as_slice());
+
+        //  4. Let output be the I/O queue of scalar values <<end-of-queue>>.
+        let mut output_buffer = [0u8; 2048];
+        //  5. While true:
+        //    1. Let item be the result of reading from this's I/O queue.
+        //    2. If item is end-of-queue and this's do not flush is true,
+        //       then return the result of running serialize I/O queue with this and output.
+        //    3. Otherwise:
+        //      1. Let result be the result of processing an item with item, this's decoder,
+        //         this's I/O queue, output, and this's error mode.
+        //      2. If result is finished, then return the result of running serialize I/O queue with this and output.
+        //      3. Otherwise, if result is error, throw a TypeError.
+
+        //  NB: the above is implemented using the encoding_rs crate:
+        //    - in the case of a chunked input, we use the decoder's decode_to* method
+        //      and we apply `self.serialize` to maintain a state in `self.io_queue`
+        //    - in the case of a plain input, we use `self.encoding.decode` on `self.io_queue`
+        // FIXME: ignore_bom should be used somewhere
+        let (result, read, written, had_errors) = self.decoder.decode_to_utf8(
+            &self.io_queue,
+            &mut output_buffer,
+            !self.do_not_flush,
+        );
+        match result {
+            // FIXME: result can indicate buffer size problem ? should we worry about that ?
```

Comment:
Yes, I think `OutputFull` could definitely occur here. 

I didn't have this context when I suggested using an array. What I suggest is switching `output_buffer`
to a `Vec<u8>` (initialized using `vec![0; 1024]`). 

In the case of `OutputFull`, we loop (using a slice of the `io_queue`)

Reply:
> In the case of OutputFull, we loop (using a slice of the io_queue)

what do you mean by using a slice of the io_queue ? we are talking about the output_buffer here isn'it it ?

Reply:
We don't want to read the same part of the IO queue each time, so we provide the slice (determined by total number of bytes read)

Reply:
Oh okay I see what you mean, I will try again

Reply:
done in the last commit